doc/developer-guide/internal-libraries/Extendible.en.rst

```diff
@@ -0,0 +1,129 @@
+   }
+
+
+When an Extendible derived class is instantiated, new() will allocate a block of memory for the derived class and all added fields. There is zero memory overhead per instance, unless using :ref:ACIDPTR field access type.
+
+.. code-block:: cpp
+
+   ExtendibleExample* alloc_example() {
+     return new ExtendibleExample();
+   }
+
+Memory Layout
+-------------
+Memory is always arranged in the following order:
+
+#. Derived members (+padding align next field)
+#. Largest Field
+#. Smallest Field
+#. Packed Bits
+
++-------------------------------------------+
+|Extendible::new()                          |
++-----------------+-------------------------+
+|ExtendibleExample|Fields                   |
++-----------------+-------------------------+
+|int real_member  |atomic<int> my_plugin_int|
++-----------------+-------------------------+
+
+Extendible was written for efficient thread safety. When a field is added to the Extendible type, data type and access type part of the addField template, resulting in a strongly typed FieldId, and then will be implicitly enforced through every access of that field. 
+
+.. code-block:: cpp
+
+   int read_my_int(ExtendibleExample &ext) {
+     return ext.get(fld_my_int);
+   }
+   void write_my_int(ExtendibleExample &ext, int val) {
+     ext.get(fld_my_int) = val;
+   }
+
+   // Common user errors
+   bool type_mismatch(ExtendibleExample &ext) {
+     return ext.writeBit(fld_my_int); // Compile error: writeBit(field<BIT,bool>): field is not of type FieldId<BIT,bool>
+   }
+   int field_mismatch(ExtendibleOtherType &ext) {
+     return ext.get(fld_my_int); // Compile error: fld_my_int is not of type ExtendibleOtherType::FieldId
+   }
+
+
+Field Access Types
+------------------
+.. _AccessType:
+
+============   =======================================   ============================================================   ========================================   ================================================
+Enums          Allocates                                 API                                                            Pros                                       Cons  
+============   =======================================   ============================================================   ========================================   ================================================
+ATOMIC         std::atomic<FieldType>                    :code:`get(FieldId)`                                           Leverages std::atomic<> API. No Locking.   Only works on small data types.
+BIT            1 bit from packed bits                    :code:`get(FieldId), readBit(FieldId), writeBit(FieldId)`      Memory efficient.                          Cannot return reference. 
+CONST          FieldType                                 :code:`get(FieldId), writeConst(FieldId)`                      Direct reference. Fast.                    No thread protection.
```

Comment:
Why would an access type of `CONST` support a `writeConst` operation? Does this mean the data is not constant?

Reply:
The CONST API is really a direct, non-thread safe access to memory. It only strongly types the FieldId and cast the return type. It is only "const" by convention. The other flaw here is that there is no obvious place to init a const field. The constructor is called, but the 

ATOMIC could be used for small type constants with acceptable overhead. 
But if large type constants (IE a const string or char[]) were to use ACIDPTR, then allocates storage elsewhere and it acquires a unique lock for each read. Which could create lock contention. 

TODO: 
1. take another look at std::shared_lock, or use atomics to allow multiple readers. 
2. take another look at providing Extendible reference to field constructors to allow constant initialization.

Reply:
> It is only "const" by convention because there is no obvious place to init a const field.

Makes sense. Might I suggest calling this something like `initializeConst`? Calling this twice would cause a runtime (supposing that we cannot enforce such a thing at compile-time).

> ATOMIC could be used for small type

Is this because it must fit into a CPU register in order to take advantage of some low-level CAS-like instruction? If so, mentioning it would help clarify what "small type" means for the dev's platform.